### PR TITLE
Add Aggregate versioning extension

### PR DESCRIPTION
**Pull request by Myks92**
Add new extension Aggregate versioning for issue #2333

### CODE REVIEW

**Review by zakharov-ilia — APPROVED** [review]

File: src/AggregateVersioning/AggregateVersionListener.php
Comment (zakharov-ilia) [inline src/AggregateVersioning/AggregateVersionListener.php:44]: I think it's better to use a array_filter() here

**Review by phansys — COMMENTED** [review]
Congratulations for your first PR @Myks92!
Is this extension compatible with the ODM?

**Comment by Myks92** [comment]
> Congratulations for your first PR @Myks92!
Thanks!

> Is this extension compatible with the ODM?
Sorry, ODM not supported. Only ORM. Also not implemented Symfony bundle. I can do it. Only I would like your review about implemented feature.

File: doc/aggregate_versioning.md
Comment (phansys) [inline doc/aggregate_versioning.md:267]:
```suggestion
     * @var Collection<int, OrderLine>
```

File: doc/aggregate_versioning.md
Comment (phansys) [inline doc/aggregate_versioning.md:310]:
```suggestion
     * @return Collection<int, OrderLine>
```

File: src/AggregateVersioning/AggregateVersionListener.php
Comment (phansys) [inline src/AggregateVersioning/AggregateVersionListener.php:67]: How many times this condition can be met?

File: src/AggregateVersioning/AggregateVersionListener.php
Comment (phansys) [inline src/AggregateVersioning/AggregateVersionListener.php:77]:
```suggestion
                throw new LogicException(sprintf('Method "%s()" not exists in class "%s".', $annotation->aggregateRootMethod, get_class($entity)));
```

File: src/AggregateVersioning/AggregateVersionListener.php
Comment (phansys) [inline src/AggregateVersioning/AggregateVersionListener.php:93]:
```suggestion
            throw new LogicException(sprintf('Aggregate "%s" must have %s.', get_class($entity), AggregateVersioning::class));
```

File: tests/Gedmo/AggregateVersioning/AggregateVersioningTest.php
Comment (phansys) [inline tests/Gedmo/AggregateVersioning/AggregateVersioningTest.php:124]:
```suggestion
        if (null === $order = $this->em->getRepository(self::ORDER)->find($id)) {
```

**Review by phansys — CHANGES_REQUESTED** [review]

File: src/AggregateVersioning/AggregateVersionListener.php
Comment (Myks92) [reply src/AggregateVersioning/AggregateVersionListener.php:65]: Сделал возврат, как только нашли агрегированный корень. https://github.com/doctrine-extensions/DoctrineExtensions/pull/2334/commits/663fff28ba20e5ebea9c0b0362edffbdce731700

File: src/AggregateVersioning/AggregateVersionListener.php
Comment (phansys) [inline src/AggregateVersioning/AggregateVersionListener.php:70]: Even if we currently know that the only method calling this one will not pass an empty array as argument 1, we should take care of that case.

**Review by phansys — CHANGES_REQUESTED** [review]

**Comment by phansys** [comment]
>Sorry, ODM not supported. Only ORM. Also not implemented Symfony bundle. I can do it. Only I would like your review about implemented feature.

I'm sorry, but as I'm not using the **Aggregate** extension, I think I'm not able to share a proper feedback about the feature.
I'm leaving comments regarding the syntax and the codebase itself.
I've requested the review from other maintainers too.

**Comment by Myks92** [comment]
> > Sorry, ODM not supported. Only ORM. Also not implemented Symfony bundle. I can do it. Only I would like your review about implemented feature.
> 
> I'm sorry, but as I'm not using the **Aggregate** extension, I think I'm not able to share a proper feedback about the feature. I'm leaving comments regarding the syntax and the codebase itself. I've requested the review from other maintainers too. Regarding the Symfony bundle, I think it is not required here, as this package does not provide a bridge for the framework. I guess the best place to bring that feature is [StofDoctrineExtensionsBundle](https://github.com/stof/StofDoctrineExtensionsBundle).

Thanks for the reviews! Made all suggested changes. I am waiting for reviews from other maintainers.

File: src/Mapping/Annotation/AggregateVersioning.php
Comment (franmomu) [inline src/Mapping/Annotation/AggregateVersioning.php:15]: IMHO we should add attribute support too, at some point support for annotation will be dropped.

**Comment by Myks92** [comment]
@phansys Hi! At what stage is PR? Does it make sense for me to continue supporting him? Will it be approved or can it be closed?

**Pull request closed without merging**
